kopf/k8s/config.py

```diff
@@ -0,0 +1,134 @@
+import logging
+from typing import Optional
+
+import pykube
+import requests
+import urllib3.exceptions
+
+logger = logging.getLogger(__name__)
+
+# Set in login(), consumed in get_pykube_cfg() and all API calls.
+_pykube_cfg: Optional[pykube.KubeConfig] = None
+
+
+class LoginError(Exception):
+    """ Raised when the operator cannot login to the cluster. """
+
+
+class AccessError(Exception):
+    """ Raised when the operator cannot access the cluster API. """
+
+
+def login():
+    """
+    Login to Kubernetes cluster, locally or remotely.
+
+    Keep the logged in state or config object in the global variables,
+    so that it can be available for future calls via the same function call.
+
+    Automatic refresh/reload of the tokens or objects also should be done here.
+    """
+
+    # Pykube login is mandatory. If it fails, the framework will not run at all.
+    login_pykube()
+
+    # We keep the official client library auto-login only because it was
+    # an implied behavior before switching to pykube -- to keep it so (implied).
+    try:
+        import kubernetes
+    except ImportError:
+        pass
+    else:
+        login_client()
+
+
+def login_pykube():
+    global _pykube_cfg
+    try:
```

Comment:
I guess we can add some "auto.." method for pykube-ng itself to not repeat the logic everywhere (I do the same as you in my projects: first try "in-cluster" then try kubeconfig file)